In [1]:
import pandas as pd
import numpy as np

# Introduction

This project is intended to provide valueable churn insight to Interconnect telecom operator using data from users of their services. Specifically, the data will be analyzed in terms of what factors contribute to Interconnect users discontinuing use of services. By the data, a binary classification machine learning model will be trained so that Interconnect can use it to predict which of their current users are likely to leave at any given time. This will allow Intercom to offer promotional codes and/or special plan options to insentivize said users to stay. The goal of this project is to maximize the model's AUC-ROC score, a metric ranging from 0 to 1 that measures how well a classification model distinguishes between classes.

# Project Plan

## Load Data

In [2]:
contract_url = 'https://raw.githubusercontent.com/pvnkd0v3/churn_prediction_final_project/refs/heads/main/contract.csv'
personal_url = 'https://raw.githubusercontent.com/pvnkd0v3/churn_prediction_final_project/refs/heads/main/personal.csv'
internet_url = 'https://raw.githubusercontent.com/pvnkd0v3/churn_prediction_final_project/refs/heads/main/internet.csv'
phone_url = 'https://raw.githubusercontent.com/pvnkd0v3/churn_prediction_final_project/refs/heads/main/phone.csv'

contract = pd.read_csv(contract_url)
personal = pd.read_csv(personal_url)
internet = pd.read_csv(internet_url)
phone = pd.read_csv(phone_url)

## View Data

In [3]:
def view(data):
    df = display(data)
    info = display(data.info())
    describe = display(data.describe())
    return df, info, describe

### Contract

In [4]:
view(contract)

customerID   BeginDate              EndDate            Type  \
0     7590-VHVEG  2020-01-01                   No  Month-to-month   
1     5575-GNVDE  2017-04-01                   No        One year   
2     3668-QPYBK  2019-10-01  2019-12-01 00:00:00  Month-to-month   
3     7795-CFOCW  2016-05-01                   No        One year   
4     9237-HQITU  2019-09-01  2019-11-01 00:00:00  Month-to-month   
...          ...         ...                  ...             ...   
7038  6840-RESVB  2018-02-01                   No        One year   
7039  2234-XADUH  2014-02-01                   No        One year   
7040  4801-JZAZL  2019-03-01                   No  Month-to-month   
7041  8361-LTMKD  2019-07-01  2019-11-01 00:00:00  Month-to-month   
7042  3186-AJIEK  2014-08-01                   No        Two year   

     PaperlessBilling              PaymentMethod  MonthlyCharges TotalCharges  
0                 Yes           Electronic check           29.85        29.85  
1                  No               Mailed check           56.95       1889.5  
2                 Yes               Mailed check           53.85       108.15  
3                  No  Bank transfer (automatic)           42.30      1840.75  
4                 Yes           Electronic check           70.70       151.65  
...               ...                        ...             ...          ...  
7038              Yes               Mailed check           84.80       1990.5  
7039              Yes    Credit card (automatic)          103.20       7362.9  
7040              Yes           Electronic check           29.60       346.45  
7041              Yes               Mailed check           74.40        306.6  
7042              Yes  Bank transfer (automatic)          105.65       6844.5  

[7043 rows x 8 columns]

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 7043 entries, 0 to 7042
Data columns (total 8 columns):
 #   Column            Non-Null Count  Dtype  
---  ------            --------------  -----  
 0   customerID        7043 non-null   object 
 1   BeginDate         7043 non-null   object 
 2   EndDate           7043 non-null   object 
 3   Type              7043 non-null   object 
 4   PaperlessBilling  7043 non-null   object 
 5   PaymentMethod     7043 non-null   object 
 6   MonthlyCharges    7043 non-null   float64
 7   TotalCharges      7043 non-null   object 
dtypes: float64(1), object(7)
memory usage: 440.3+ KB


None

MonthlyCharges
count     7043.000000
mean        64.761692
std         30.090047
min         18.250000
25%         35.500000
50%         70.350000
75%         89.850000
max        118.750000

(None, None, None)

In [5]:
print('Number of unique customer IDs:', contract['customerID'].nunique())
print('Proportion of active users:', len(contract[contract['EndDate'] == 'No']) / len(contract))

Number of unique customer IDs: 7043
Proportion of active users: 0.7346301292063041


**The data includes the unique customer IDs of 7043 Interconnect users and is unbalanced with the majority of users (73.5%) having not yet left.**

**Preprocessing Tasks**
- Convert column names to snakecase
- Convert BeginDate to date type
- Convert TotalCharges to float64
- Check for grammatical errors among unique values of object type columns 
- Check for inconspicuous missing values and impute as needed

### Personal

In [6]:
view(personal)

customerID  gender  SeniorCitizen Partner Dependents
0     7590-VHVEG  Female              0     Yes         No
1     5575-GNVDE    Male              0      No         No
2     3668-QPYBK    Male              0      No         No
3     7795-CFOCW    Male              0      No         No
4     9237-HQITU  Female              0      No         No
...          ...     ...            ...     ...        ...
7038  6840-RESVB    Male              0     Yes        Yes
7039  2234-XADUH  Female              0     Yes        Yes
7040  4801-JZAZL  Female              0     Yes        Yes
7041  8361-LTMKD    Male              1     Yes         No
7042  3186-AJIEK    Male              0      No         No

[7043 rows x 5 columns]

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 7043 entries, 0 to 7042
Data columns (total 5 columns):
 #   Column         Non-Null Count  Dtype 
---  ------         --------------  ----- 
 0   customerID     7043 non-null   object
 1   gender         7043 non-null   object
 2   SeniorCitizen  7043 non-null   int64 
 3   Partner        7043 non-null   object
 4   Dependents     7043 non-null   object
dtypes: int64(1), object(4)
memory usage: 275.2+ KB


None

SeniorCitizen
count    7043.000000
mean        0.162147
std         0.368612
min         0.000000
25%         0.000000
50%         0.000000
75%         0.000000
max         1.000000

(None, None, None)

In [7]:
print('Number of unique customer IDs:', personal['customerID'].nunique())

Number of unique customer IDs: 7043


**The data includes the unique customer ids of 7043 Interconnect users.**

**Preprocessing Tasks**
- Convert column names to snakecase
- Check for grammatical errors among unique values of object type columns 
- Check for inconspicuous missing values and impute as needed

###  Internet

In [8]:
view(internet)

customerID InternetService OnlineSecurity OnlineBackup DeviceProtection  \
0     7590-VHVEG             DSL             No          Yes               No   
1     5575-GNVDE             DSL            Yes           No              Yes   
2     3668-QPYBK             DSL            Yes          Yes               No   
3     7795-CFOCW             DSL            Yes           No              Yes   
4     9237-HQITU     Fiber optic             No           No               No   
...          ...             ...            ...          ...              ...   
5512  6840-RESVB             DSL            Yes           No              Yes   
5513  2234-XADUH     Fiber optic             No          Yes              Yes   
5514  4801-JZAZL             DSL            Yes           No               No   
5515  8361-LTMKD     Fiber optic             No           No               No   
5516  3186-AJIEK     Fiber optic            Yes           No              Yes   

     TechSupport StreamingTV StreamingMovies  
0             No          No              No  
1             No          No              No  
2             No          No              No  
3            Yes          No              No  
4             No          No              No  
...          ...         ...             ...  
5512         Yes         Yes             Yes  
5513          No         Yes             Yes  
5514          No          No              No  
5515          No          No              No  
5516         Yes         Yes             Yes  

[5517 rows x 8 columns]

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 5517 entries, 0 to 5516
Data columns (total 8 columns):
 #   Column            Non-Null Count  Dtype 
---  ------            --------------  ----- 
 0   customerID        5517 non-null   object
 1   InternetService   5517 non-null   object
 2   OnlineSecurity    5517 non-null   object
 3   OnlineBackup      5517 non-null   object
 4   DeviceProtection  5517 non-null   object
 5   TechSupport       5517 non-null   object
 6   StreamingTV       5517 non-null   object
 7   StreamingMovies   5517 non-null   object
dtypes: object(8)
memory usage: 344.9+ KB


None

customerID InternetService OnlineSecurity OnlineBackup  \
count         5517            5517           5517         5517   
unique        5517               2              2            2   
top     4310-KEDTB     Fiber optic             No           No   
freq             1            3096           3498         3088   

       DeviceProtection TechSupport StreamingTV StreamingMovies  
count              5517        5517        5517            5517  
unique                2           2           2               2  
top                  No          No          No              No  
freq               3095        3473        2810            2785

(None, None, None)

In [9]:
print('Number of unique customer IDs:', internet['customerID'].nunique())

Number of unique customer IDs: 5517


**The data includes the unique IDs of 5517 Interconnect customers who use internet services.**

**Preprocessing Data**
- Convert column names to snake case
- Check for grammatical errors among unique values of object type columns 
- Check for inconspicuous missing values and impute as needed

### Phone

In [10]:
view(phone)

customerID MultipleLines
0     5575-GNVDE            No
1     3668-QPYBK            No
2     9237-HQITU            No
3     9305-CDSKC           Yes
4     1452-KIOVK           Yes
...          ...           ...
6356  2569-WGERO            No
6357  6840-RESVB           Yes
6358  2234-XADUH           Yes
6359  8361-LTMKD           Yes
6360  3186-AJIEK            No

[6361 rows x 2 columns]

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 6361 entries, 0 to 6360
Data columns (total 2 columns):
 #   Column         Non-Null Count  Dtype 
---  ------         --------------  ----- 
 0   customerID     6361 non-null   object
 1   MultipleLines  6361 non-null   object
dtypes: object(2)
memory usage: 99.5+ KB


None

customerID MultipleLines
count         6361          6361
unique        6361             2
top     4979-HPRFL            No
freq             1          3390

(None, None, None)

In [11]:
print('Number of unique customer IDs:', phone['customerID'].nunique())

Number of unique customer IDs: 6361


**The data includes the unique IDs of 6361 Interconnect customers who use phone services.**

**Preprocessing Tasks**
- Convert column names to snakecase
- Check for grammatical errors among unique values of MultipleLines
- Check for inconspicuous missing values and impute as needed

## Proposed Work Plan

- Complete the preprocessing tasks listed above for each dataset
- Merge all four datasets on customerID and set said column as the index
- Perform exploratory data analysis by providing descriptive statistics and visuals on relationships between features and if users have left Interconnect or not (Whether phone or internet customers are more likely to leave, if there's seasonality amongst beginning and end dates, what features of phone and internet plans are correlated with churn, personal info of users who have left vs not, etc.)
- Engineer features by encoding non-numerical data, performing a train/test split, and scaling data to address class imbalance
- Train several different kinds of classification models using a function that trains a given model and tunes it to optimize AUC-ROC score
- Train model with optimal parameters/AUC-ROC score on both training and validation data
- Obtain final AUC-ROC score from test dataset
- Write a report on the project's findings and results

<div class="alert alert-block alert-success">
<b>Reviewer's comment</b> This looks like a good plan! Best of luck.
    <a class="tocSkip"></a>